pkg/chrootarchive: fix FreeBSD build

akhramov opened the pull request:
**- What I did**
Added a stub-fix for pkg/chrootarchive to make it build on FreeBSD.


**- How I did it**

Added FreeBSD-specific chrooted tar/untar implementation

**- How to verify it**

The problem was discovered by @crazy-max : https://github.com/moby/buildkit/pull/2376#issuecomment-1579515762

With the proposed patch, buildkit port builds and functions fine. Also the tests that have been written for linux pass.

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

cpuguy83 commented:
It seems like putting these stubs in defeats the purpose of this package.
I'm inclined to 👎 this especially since freebsd actually supports chroots.

akhramov commented:
No worries, thanks for the feedback! 

I will create a proper solution with reexec + chroot (maybe even jailing). Moving this to draft.

akhramov committed:
pkg/chrootarchive: fix FreeBSD build

For unix targets, `goInChroot()` is only implemented for `Linux`,
hence FreeBSD build fails.

This change

- Adds FreeBSD-specific chrooted tar/untar implementation
- Fixes statUnix() to accomodate to FreeBSD devminor/devmajor
- quirk. See also: https://github.com/containerd/containerd/pull/5991

Signed-off-by: Artem Khramov <akhramov@pm.me>
Co-authored-by: Cory Snider <corhere@gmail.com>

corhere reviewed (changes requested):
I'm sad to see that reexec has to stick around for non-Linux platforms, but I couldn't find any alternatives either. FreeBSD just doesn't let you play nearly as fast and loose with threads as Linux. `pkg/reexec` itself could use some attention as it currently uses the naive argv[0] method to locate the executable path on *BSD. The Go standard library has added [`os.Executable()`](https://pkg.go.dev/os#Executable) since that package was written, which knows how to use `sysctl(3)` on FreeBSD to reliably look up the path.

thaJeztah commented on pkg/chrootarchive/archive_freebsd.go:
```diff
@@ -0,0 +1,237 @@
+	defer optionsPipe.Close()
+	options, err := decodeOptions(optionsPipe)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, errors.Wrap(err, "could not decode tar options"))
```
Very quick blurb (about to head into a call); I think all of these errors (also elsewhere in this file) are part of the  reexec's and won't be unwrapped by anyone, other than being printed on the command-line.

I would suggest using stdlib  `error` / `fmt.Errorf()` for these if we don't need the added features of pkg/errors.

For this specific case, I guess it doesn't even need to be an error, as we're just printing, so this could be a `fmt.Fprintf` (with a newline in the format string)

akhramov replied:
Thanks, fixed.

akhramov commented on pkg/chrootarchive/archive_freebsd.go:
```diff
@@ -0,0 +1,237 @@
+	var options archive.TarOptions
+
+	return options, decodeJson(options, reader)
+}
```
Agreeed, went ahead and inlined these.

akhramov commented on pkg/chrootarchive/archive_freebsd.go:
```diff
@@ -0,0 +1,237 @@
+	return done, nil
+}
+
+func unpackInChroot() {
```
It's a really great idea that allowed me to drastically simplify the entrypoints, thank you!

akhramov commented on pkg/chrootarchive/archive_freebsd.go:
```diff
@@ -0,0 +1,237 @@
+	return r, nil
+}
+
+func doUnpackLayer(root string, layer io.Reader, options *archive.TarOptions) (chan (unpackLayerResult), error) {
```
Fixed, thanks!

akhramov commented on pkg/archive/archive.go:
```diff
@@ -53,30 +53,30 @@ type (
+		NoLchown         bool                    `json:"NoLchown"`
+		IDMap            idtools.IdentityMapping `json:"IDMap"`
+		ChownOpts        *idtools.Identity       `json:"ChownOpts"`
+		IncludeSourceDir bool                    `json:"IncludeSourceDir"`
```
Thanks for the clarification, I learned a lot today. 

I went ahead and used gob instead of json which sounds more reasonable for IPC use-case.

corhere commented on pkg/chrootarchive/archive_test.go:
```diff
@@ -35,6 +36,13 @@ func CopyWithTar(src, dst string) error {
 	return chrootArchiver.CopyWithTar(src, dst)
 }
 
+func TestMain(m *testing.M) {
+	if reexec.Init() {
+		return
+	}
+	m.Run()
+}
```
I would prefer that `TestMain` be moved to a new file `archive_freebsd_test.go` so that the reexec package isn't linked into the tests on other platforms.

akhramov replied:
Sure, thanks!

corhere reviewed (changes requested):
Just a couple of little things to address and I think we're good to go!

cpuguy83 commented on pkg/chrootarchive/archive_freebsd.go:
```diff
@@ -0,0 +1,213 @@
+				"could not chroot: %v\n",
+				os.PathError{Op: "chroot", Path: root, Err: err},
+			)
+			os.Exit(1)
```
Can you increment the exit codes here and at L50?
This way exit code 1 is caller error, 2 is chroot error, and 3 is operation error.

akhramov replied:
Makes sense, thanks!

cpuguy83 reviewed (commented):
Minor comment and +1 to all of @corhere's
Otherwise looks good.

akhramov commented on pkg/chrootarchive/archive_freebsd.go:
```diff
@@ -0,0 +1,213 @@
+
+	io.Copy(os.Stdout, tb.Reader())
+
+	return nil
```
Thanks, fixed!

akhramov commented on pkg/chrootarchive/archive_freebsd.go:
```diff
@@ -0,0 +1,213 @@
+func doUnpack(decompressedArchive io.Reader, relDest, root string, options *archive.TarOptions) error {
+	optionsR, optionsW, err := os.Pipe()
+	if err != nil {
+		return &os.SyscallError{Syscall: "pipe", Err: err}
```
Thanks!

akhramov commented on pkg/chrootarchive/archive_freebsd.go:
```diff
@@ -0,0 +1,213 @@
+				os.Stderr,
+				"could not chroot: %v\n",
+				os.PathError{Op: "chroot", Path: root, Err: err},
+			)
```
Thanks, fixed!

akhramov commented on pkg/chrootarchive/archive_freebsd.go:
```diff
@@ -0,0 +1,213 @@
+func reexecMain(f func(options archive.TarOptions, args ...string) error) func() {
+	return func() {
+		options, err := recvOptions()
+		root := os.Args[1]
```
Sure thing, fixed

akhramov commented on pkg/chrootarchive/archive_freebsd.go:
```diff
@@ -0,0 +1,213 @@
+
+import (
+	"bytes"
+	"encoding/gob"
```
That makes perfect sense, thank you!

corhere commented on pkg/chrootarchive/archive_freebsd.go:
```diff
@@ -0,0 +1,225 @@
+func doPack(relSrc, root string, options *archive.TarOptions) (io.ReadCloser, error) {
+	optionsR, optionsW, err := os.Pipe()
+	if err != nil {
+		return nil, &os.SyscallError{Syscall: "pipe", Err: err}
```
```suggestion
		return nil, err
```

corhere commented on pkg/chrootarchive/archive_freebsd.go:
```diff
@@ -0,0 +1,225 @@
+	cmd.Stderr = stderr
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, &os.SyscallError{Syscall: "pipe", Err: err}
```
```suggestion
		return nil, err
```

corhere commented on pkg/chrootarchive/archive_freebsd.go:
```diff
@@ -0,0 +1,225 @@
+	var result int64
+	optionsR, optionsW, err := os.Pipe()
+	if err != nil {
+		return 0, &os.SyscallError{Syscall: "pipe", Err: err}
```
```suggestion
		return 0, err
```

corhere reviewed (changes requested):
Only wrap errors returned from `syscall` or `golang.org/x/sys/unix` functions (i.e. errors of type [`syscall.Errno`](https://pkg.go.dev/syscall#Errno)) in `os.SyscallError` or `fs.PathError`.

corhere reviewed (approved):
LGTM, thanks!
